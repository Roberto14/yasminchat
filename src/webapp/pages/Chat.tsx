import React from 'react';
import { Container } from 'theme-ui';
import MessageInput from '../components/MessageInput';
import MessagesList from '../components/MessagesList';
import { MessagesConsumer } from '../providers/MessagesProvider';
import Navigation from '../components/Navigation';

const Chat = () => (
  <MessagesConsumer>
    {({ messages, onSend, read }) => {
      const gridTemplateRows = ['3em 1fr 5em', '3em 1fr 6em', '3em 1fr 8em'];

      return (
        <Container variant="body" sx={{ gridTemplateRows }}>
          <Navigation />
          <Container variant="content" sx={{ overflowY: 'scroll' }}>
            <MessagesList messages={messages} read={read} />
          </Container>
          <Container variant="content">
            <MessageInput onSend={onSend} />
          </Container>
        </Container>
      );
    }}
  </MessagesConsumer>
);

export default Chat;
