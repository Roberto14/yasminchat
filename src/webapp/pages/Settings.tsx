import React, { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import {
  useColorMode, Button, Label, Input, Box, Container,
} from 'theme-ui';
import useStateWithLocalStorage from '../hooks/useStateWithLocalStorage';
import {
  SETTING_ENTER_ENABLED, SETTING_USER, SETTING_TIME_FORMAT_24,
  DEFAULT_ENTER_ENABLED, DEFAULT_USER, DEFAULT_INTERFACE_MODE, DEFAULT_TIME_FORMAT_24,
  LIGHT, DARK, ON, OFF, DEFAULT_LANG,
} from '../constants';
import BinaryRadioInput from '../components/BinaryRadioInput';
import LocaleSwitcher from '../components/LocaleSwitcher';
import Navigation from '../components/Navigation';

const Settings = () => {
  // Load all settings from LocalStorage
  const [user, setUser] = useStateWithLocalStorage(SETTING_USER);
  const [timeFormat24, setTimeFormat24] = useStateWithLocalStorage(SETTING_TIME_FORMAT_24);
  const [enterEnabled, setEnterEnabled] = useStateWithLocalStorage(SETTING_ENTER_ENABLED);
  const [interfaceMode, setInterfaceMode] = useColorMode();
  // can load namespace to access more specific set of translations
  const { i18n, t } = useTranslation();

  // callback function to set all defaults
  // useCallback with no dependencies array, this will keep the same function ref
  // as we dont need to re-defined it
  const setDefaults = useCallback(() => {
    setUser(DEFAULT_USER);
    setInterfaceMode(DEFAULT_INTERFACE_MODE);
    setTimeFormat24(DEFAULT_TIME_FORMAT_24);
    setEnterEnabled(DEFAULT_ENTER_ENABLED);
    i18n.changeLanguage(DEFAULT_LANG);
  }, []);

  const mt = 4; // marginTop for each row

  return (
    <Container variant="body" sx={{ gridTemplateRows: '3em 1fr 3em' }}>
      <Navigation />

      <Container variant="content" sx={{ overflowY: 'scroll' }}>

        <Box>
          <Label>{t('Username')}</Label>
          <Input mb={mt} value={user} onChange={useCallback((e) => setUser(e.target.value), [])} />
        </Box>

        <Box mt={mt}>
          <Label mb={2}>{t('Interface Colour')}</Label>
          <BinaryRadioInput
            name="interfaceMode"
            options={[
              { text: 'Light', value: LIGHT, selected: interfaceMode === LIGHT },
              { text: 'Dark', value: DARK, selected: interfaceMode === DARK },
            ]}
            onChange={useCallback((e) => setInterfaceMode(e.target.value), [])}
          />
        </Box>

        <Box mt={mt}>
          <Label mb={2}>{t('Clock Display')}</Label>
          <BinaryRadioInput
            name="clockDisplay"
            options={[
              { text: '12 Hours', value: OFF, selected: timeFormat24 === OFF },
              { text: '24 Hours', value: ON, selected: timeFormat24 === ON },
            ]}
            onChange={useCallback((e) => setTimeFormat24(e.target.value), [])}
          />
        </Box>

        <Box mt={mt}>
          <Label mb={1}>{t('Send Messages on CTRL + ENTER')}</Label>
          <BinaryRadioInput
            name="ctrlEnter"
            options={[
              { text: 'On', value: ON, selected: enterEnabled === ON },
              { text: 'Off', value: OFF, selected: enterEnabled === OFF },
            ]}
            onChange={useCallback((e) => setEnterEnabled(e.target.value), [])}
          />
        </Box>

        <Box mt={mt}>
          <Label>{t('Language')}</Label>
          <LocaleSwitcher />
        </Box>

      </Container>

      <Container variant="content">
        <Button id="reset" sx={{ width: '100%' }} onClick={setDefaults}>{t('Reset do Defaults')}</Button>
      </Container>

    </Container>
  );
};

export default Settings;
