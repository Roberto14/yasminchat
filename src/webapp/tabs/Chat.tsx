import React from 'react'
import MessageInput from "../components/MessageInput"
import MessagesList from "../components/MessagesList"
import { MessagesConsumer } from "../providers/MessagesProvider"

const Chat = () => (
    <MessagesConsumer>
        {({messages, onSend, read }) => {
            // bit hacky but still quick and clean way to avoid MessagesProvider to setState during render
            setTimeout(read)

            return (
                <div>
                    <MessagesList messages={messages} />
                    <MessageInput onSend={onSend} />
                </div>
            )
        }}
    </MessagesConsumer>
)

export default Chat
