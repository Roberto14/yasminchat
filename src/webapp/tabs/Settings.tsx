import React, {useCallback} from 'react';
import useStateWithLocalStorage from "../hooks/useStateWithLocalStorage"
import {
    SETTING_ENTER_ENABLED, SETTING_USER, SETTING_INTERFACE_MODE, SETTING_TIME_FORMAT_24,
    DEFAULT_ENTER_ENABLED, DEFAULT_USER, DEFAULT_INTERFACE_MODE, DEFAULT_TIME_FORMAT_24,
    LIGHT, DARK, ON, OFF,
} from "../../constants";
import BinaryRadioInput from "../components/BinaryRadioInput";
import { useTranslation } from 'react-i18next';
import LocaleSwitcher from "../components/LocaleSwitcher";

const Settings = () => {

    const [user, setUser] = useStateWithLocalStorage(SETTING_USER)
    const [interfaceMode, setInterfaceMode] = useStateWithLocalStorage(SETTING_INTERFACE_MODE)
    const [timeFormat24, setTimeFormat24] = useStateWithLocalStorage(SETTING_TIME_FORMAT_24)
    const [enterEnabled, setEnterEnabled] = useStateWithLocalStorage(SETTING_ENTER_ENABLED)

    const setDefaults = useCallback(() =>{
        setUser(DEFAULT_USER)
        setInterfaceMode(DEFAULT_INTERFACE_MODE)
        setTimeFormat24(DEFAULT_TIME_FORMAT_24)
        setEnterEnabled(DEFAULT_ENTER_ENABLED)
    }, [])

    const { t } = useTranslation() // can load namespace to access more specific set of translations

    return (
        <div>
            <div>
                <p>{t('Username')}</p>
                <input type="text" value={user} onChange={(e) => setUser(e.target.value)}/>
            </div>

            <div>
                <p>{t('Interface Colour')}</p>
                <BinaryRadioInput
                    name="interfaceMode"
                    options={[
                        { text: 'Light', value: LIGHT, selected: interfaceMode === LIGHT},
                        { text: 'Dark', value: DARK, selected: interfaceMode === DARK}
                    ]}
                    onChange={(e) => setInterfaceMode(e.target.value)}
                />
            </div>

            <div>
                <p>{t('Clock Display')}</p>
                <BinaryRadioInput
                    name="clockDisplay"
                    options={[
                        { text: '12 Hours', value: OFF, selected: timeFormat24 === OFF},
                        { text: '24 Hours', value: ON, selected: timeFormat24 === ON}
                    ]}
                    onChange={(e) => setTimeFormat24(e.target.value)}
                />
            </div>


            <div>
                <p>{t('Send Messages on CTRL + ENTER')}</p>
                <BinaryRadioInput
                    name="ctrlEnter"
                    options={[
                        { text: 'On', value: ON, selected: enterEnabled === ON},
                        { text: 'Off', value: OFF, selected: enterEnabled === OFF}
                    ]}
                    onChange={(e) => setEnterEnabled(e.target.value)}
                />
            </div>

            <div>
                <p>{t('Language')}</p>
                <LocaleSwitcher />
            </div>

            <button type="button" onClick={() => setDefaults()}>{t('Reset do Defaults')}</button>

        </div>
    )
}

export default Settings
