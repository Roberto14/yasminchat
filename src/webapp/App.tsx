import React from "react"
import Chat from './pages/Chat'
import Settings from './pages/Settings'
import { loadSettingsOrDefaults } from './helpers/loadSettings'
import { MessagesProvider } from "./providers/MessagesProvider"
import { HashRouter, Switch, Route } from "react-router-dom"
import { ThemeProvider } from 'theme-ui'
import theme from './theme'
import io from 'socket.io-client'

// App class component loads the essential dependencies we need to make the app work
// It will load: socketio, router, theme-ui provider and MessagesProvider

type AppStateType = {
    loading: boolean,   // we could use this property to show a spinner while the app is not ready
}

class App extends React.Component<{}, AppStateType> {
    state: AppStateType = {
        loading: true,
    }

    socket: null | SocketIOClient.Socket = null

    componentDidMount() {
        // connect to socket io
        this.socket = io('http://localhost:3000')

        // Load localstorage settings or set defaults on first time
        loadSettingsOrDefaults()

        // mutate state to let the page render
        this.setState({ loading: false})
    }

    render() {
        const { loading } = this.state
        const { socket } = this
        return (
            <HashRouter>
                <ThemeProvider theme={theme}>
                    {!loading && socket && (
                        <MessagesProvider socket={socket}>
                            <Switch>
                                <Route path="/settings">
                                    <Settings />
                                </Route>
                                <Route path="/">
                                    <Chat />
                                </Route>
                            </Switch>
                        </MessagesProvider>
                    )}
                </ThemeProvider>
            </HashRouter>
        )
    }
}

export default App
