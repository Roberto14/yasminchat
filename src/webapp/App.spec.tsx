import React from 'react';
import { shallow } from 'enzyme'
import App from './App'
import Chat from "./pages/Chat";
import Settings from "./pages/Settings";
import socketIOClient from 'socket.io-client'
// @ts-ignore
import MockedSocket from 'socket.io-mock'

jest.mock('socket.io-client')

describe('App Component', ()=>{
    beforeEach(() => {
        // @ts-ignore
        socketIOClient.mockReturnValue(new MockedSocket());
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('matches the snapshot', () => {
        const wrapper = shallow(<App />)
        expect(wrapper).toMatchSnapshot()
    })

    it('checks socket.io has been loaded and called with correct url', () => {
        shallow(<App />) // only needs to run the wrapper
        const url = 'http://localhost:3000'

        expect(socketIOClient.connect).toHaveBeenCalledWith(url);
    })

    // This tests is not entirely needed, snapshot matching is sufficient
    it('should find both Chat and Settings page components', () =>{
        const wrapper = shallow(<App />)

        expect(wrapper.find(Chat)).toHaveLength(1)
        expect(wrapper.find(Settings)).toHaveLength(1)
    })
})
