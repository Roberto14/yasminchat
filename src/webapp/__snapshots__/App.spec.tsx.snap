// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches the snapshot 1`] = `
<HashRouter>
  <ThemeProvider
    theme={
      Object {
        "breakpoints": Array [
          "768px",
          "1024px",
          "1200px",
        ],
        "buttons": Object {
          "primary": Object {
            "&:hover": Object {
              "bg": Array [
                "none",
                "none",
                "text",
              ],
            },
            "bg": "primary",
            "color": "background",
            "cursor": "pointer",
            "outline": 0,
          },
        },
        "colors": Object {
          "background": "#fff",
          "messages": "#1a202c",
          "modes": Object {
            "dark": Object {
              "background": "#1a202c",
              "messages": "#1a202c",
              "notifications": "#2a94e0",
              "primary": "#ccc",
              "text": "#fff",
            },
          },
          "notifications": "#cc0000",
          "primary": "#07c",
          "text": "#1a202c",
        },
        "fontSizes": Array [
          12,
          14,
          16,
          20,
          24,
          32,
        ],
        "fonts": Object {
          "body": "system-ui, sans-serif",
        },
        "forms": Object {
          "input": Object {
            "&:focus": Object {
              "borderColor": "primary",
            },
            "borderColor": "gray",
            "maxWidth": Array [
              "100%",
              "30em",
            ],
            "outline": 0,
          },
          "label": Object {
            "fontSize": Array [
              1,
              2,
              2,
            ],
            "fontWeight": "bold",
          },
          "select": Object {
            "&:focus": Object {
              "borderColor": "primary",
            },
            "borderColor": "gray",
            "maxWidth": Array [
              "100%",
              "30em",
            ],
            "outline": 0,
          },
          "textarea": Object {
            "&:focus": Object {
              "borderColor": "primary",
            },
            "borderColor": "gray",
            "fontSize": Array [
              0,
              0,
              2,
            ],
            "height": Array [
              "4em",
              "5em",
            ],
            "outline": 0,
            "resize": "none",
          },
        },
        "layout": Object {
          "body": Object {
            "display": "grid",
            "height": "100vh",
            "padding": "0 1em",
          },
          "content": Object {
            "mt": 2,
            "p": Array [
              0,
              0,
              "0 20%",
            ],
          },
        },
        "messages": Object {
          "received": Object {
            "background": "#efeffe",
            "color": "messages",
            "float": "left",
          },
          "sent": Object {
            "background": "#e4f6e2",
            "borderLeft": "none",
            "borderRight": "solid 4px",
            "borderRightColor": "green",
            "color": "messages",
            "float": "right",
          },
        },
        "styles": Object {
          "root": Object {
            "fontFamily": "body",
            "height": "100%",
            "m": 0,
          },
        },
      }
    }
  >
    <MessagesProvider
      socket={
        Object {
          "_emitFn": [Function],
          "broadcast": Object {
            "to": [Function],
          },
          "generalCallbacks": Object {},
          "joinedRooms": Array [],
          "rooms": Array [],
          "socketClient": e {
            "_emitFn": [Function],
            "_socketMock": [Circular],
          },
        }
      }
    >
      <Switch>
        <Route
          path="/settings"
        >
          <Settings />
        </Route>
        <Route
          path="/"
        >
          <Chat />
        </Route>
      </Switch>
    </MessagesProvider>
  </ThemeProvider>
</HashRouter>
`;
